### PR TITLE
Sanitize posts before plugin filters, not afterward

### PR DESCRIPTION
**Pull request by LukeLaupheimer**
In [this thread](https://community.nodebb.org/topic/9057/translator-ignoring-entries-that-don-t-pass-a-parameter-to-it/), I raised a question about why parameterless keys (ex, ``[[subject:key]]` as opposed to `[[subject:key, blah]]`) were not being evaluated by the translator when the source of these keys is a plugin modifying the post's content after injecting it with a partial. See my [New World of Darkness Dice Roller](https://github.com/LukeLaupheimer/nodebb-plugin-nwod-roller/blob/c54dbbe4182f2bc6ac6c1007e4f717827a4c7414/library.js)'s implementation of that on line 33. A "hacky" solution that puts in a blank parameter for the key is included in the linked plugin's most recent commit.

After researching why this takes place, it turns out that the actual source of the issue was not in the translator's code, but in the `parsePost` method. It would sanitize inputs *after* letting the plugins run their course, not *before*-hand. Sanitization of this type should always be done prior to programmatic manipulation of data. A test to this effect is included in this pull request.

### CODE REVIEW

**Comment by LukeLaupheimer** [comment]
A friend of mine has told me this pull request reads as condescending/patronizing. I didn't mean it that way at all. I just felt I had to document why I'm doing what I'm doing thoroughly, since I'm going into someone else's world and changing around their code. :)

File: tests/posts.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
+'use strict';
```
Comment (julianlam) [inline tests/posts.js]: Wow, tests :smile:

**Comment by julianlam** [comment]
Thanks @LukeLaupheimer -- unfortunately I cannot determine whether this is good because `translator.escape()` was a method introduced by @psychobunny, so I would prefer to have him look this over.

**Comment by enable-labs** [comment]
That would make sense that we get his input. @psychobunny: the tl;dr version of the OP is that this change makes `Posts.parsePost()` sanitize prior to the `filter:parse.post` hook, so that it sanitizes *user input* but not *plugin manipulation*.

**Comment by LukeLaupheimer** [comment]
That makes sense, @julianlam to have @psychobunny's input. That said, the code that actually gets changed is not `translator.escape()`, but `Posts.parsePost()`. The tl;dr version of the whole point of this is to make it so that it sanitizes the code prior to running the filter hook, so that it only sanitizes user input, as opposed to plugin manipulations of that input.

**Comment by akhoury** [comment]
I like the idea, but there is nothing stoping a malicious plugin of `unescaping` back the content, right? (or doing any other damage for that matter)

Malicious plugins cannot really be avoided, well maybe some rating and flagging system.

**Comment by LukeLaupheimer** [comment]
I have three reasons why I don't think this is a big deal, in no particular order:
* The damage that a malicious plugin can do in this context is allow users to use keys to insert system messages, which is at worst humorous if it's done in the middle of a post, and
* Doing this doesn't cause any permanent damage and can be corrected by simply disabling the plugin,
* Plugins being able to do localization on the server side *in media res* on a post to me seems like an important part of being able to have plugins manipulate post content to begin with.

**Comment by psychobunny** [comment]
I think this is okay, you guys have my signoff but probably @barisusakli should look at it since he's the last person to have done much damage (I meant refactoring) to translator.js 


(... hot :sweet_potato: 'd)

**Comment by barisusakli** [comment]
Unfortunately this seems to break https://github.com/NodeBB/NodeBB/issues/4362

**Comment by LukeLaupheimer** [comment]
It doesn't. If a user inputs `[[xy:z]]`, it filters that. However, if a plugin adds `[[xy:z]]` after the fact, it does not. This is tested for on `./tests/posts.js:10`.

**Comment by barisusakli** [comment]
Well I moved `postData.content = translator.escape(postData.content);` before the hook and when someone posts `[[global:home]]` as post content it gets translated, which doesn't happen when it is after the hook.

**Comment by LukeLaupheimer** [comment]
Oh wow, you're right. I just tried it. Sorry about that mistake on my part.

I guess either take it as-is and let it override #4362, or close this pull request without merging it, give me some time, and I'll figure out why it's `unescape`ing this after the fact. Because that's very curious.

**Comment by LukeLaupheimer** [comment]
Hi guys. I found the issue. As the tests in the file show, it does parse it correctly. However, one of the plugins that is installed by default does cause it to essentially unescape. In @julianlam's markdown plugin, on `index.js:107`, you get this line:
```javascript
data.postData.content = parser.render(data.postData.content);
```

By no means do I think we should get rid of this. However, what I'm going to spend some free time tonight is looking into how to escape the keys without tripping up the Markdown interpreter. And hopefully, without accidentally causing other issues along the way.

**Comment by LukeLaupheimer** [comment]
This commit escapes the localization by placing a zero-width white space character in between each square braces pair. In my own opinion, this *looks* a bit hacky but it also does address the issue that @barisusakli raised, and it's a very valid one.

**Comment by LukeLaupheimer** [comment]
So since I was altering `translator.escape()`, I figured there was a possibility that I might shoot myself in the foot by accident. and can some unforeseen consequences. So I ran it on my private forum that I run, and lo' and behold, my users dutifully pointed out that the top menu bar went out of whack. Turns out, some things are hard-coded as escaped in NodeBB, which I didn't know until then.

So I restored the old `translator.escape()` code and broke my approach out into its own thing. Two functions: `translator.sanitizeHTML()` and `translator.unsanitizeHTML()` (not yet used).

**Comment by LukeLaupheimer** [comment]
Are there any other issues you have with this, @barisusakli?

**Comment by barisusakli** [comment]
@LukeLaupheimer in your plugin instead of using app.render here https://github.com/LukeLaupheimer/nodebb-plugin-nwod-roller/blob/c54dbbe4182f2bc6ac6c1007e4f717827a4c7414/library.js#L46 can you instead use 
`templates.parse` you can require the templates module from nodebb itself with `var templates = module.parent.require('templates.js');` This might remove the need for any changes to core.

What you are doing right now is basically rendering a partial using `app.render` which uses both templates.js and the translator and then you are injecting the resulting html into the postData.content which then goes into app.render to be rendered again. I think this is what is causing the issue in the first place.

Your `Roller.rollHTML` function should look like below.

```
Roller.rollHTML = function(query, results, again, pool, callback) {
  var renderData = {
    "pool" : pool,
    "query" : query,
    "results" : results,
    "again" : again,
    "differentAgain" : (again != 10),
    "noRerolling" : (again > 10),
    "results" : results.join(" "),
    "noRerollingPlaceholder" : "",
    "successes" : Roller.successes(results)
  };

  templates.parse("partials/nwod-roll", renderData, function (err, html) {
    callback(html);
  });
};
```

**Comment by LukeLaupheimer** [comment]
@barisusakli I gave that a shot, and it didn't work. It just returned "undefined" where the outcome would have been. I can fiddle around with it some more later, but it definitely won't work without some tweaking, somewhere, and I haven't figured out where yet.

Also, [the documentation recommends that I use `app.render()`](https://nodebb.readthedocs.io/en/latest/themes/templates.html), so I don't think I'm doing anything incorrect in this regard. I've also already created a [functioning workaround](https://github.com/LukeLaupheimer/nodebb-plugin-nwod-roller/blob/c54dbbe4182f2bc6ac6c1007e4f717827a4c7414/library.js#L42) by just inserting a blank parameter into the key. Works like a charm.

This isn't about me getting my plugin to function; I've already accomplished that with the current code. This is about correcting a real and actual bug in NodeBB's code. That bug being it sanitizes the post after plugin filters fire, and in doing so eliminates the possibility of using partials in filter hooks. If you think this is how the software ought to work, just close the pull request.

**Comment by julianlam** [comment]
With templates.parse, there is only one argument in the callback, the output HTML, no error

**Comment by barisusakli** [comment]
@LukeLaupheimer can you switch back to using app.render and let me know the value of `html` from the app.render call [here](https://github.com/LukeLaupheimer/nodebb-plugin-nwod-roller/blob/master/library.js#L59)

Since your partial has the `[[nwodroll:results]]`` it should be translated and inserted into the `postData.content` value so I don't see why it wouldn't work unless there is a bug in the translator itself.

Here is how the flow goes as far as I can tell.

1. Posts.parsePost is called
2. `filter:parse.post` is fired and passed the post
3. Your plugin renders(templates.parse & translates) the partial and inserts it into postData.content
    At this point the postData.content should have the translated value of `[[nwodroll:results]]`
4. Back in `Posts.parse` calling `translator.escape` is called, but this shouldn't effect you since your key should already be translated.

**Comment by barisusakli** [comment]
@LukeLaupheimer I sent your plugin a PR, turns out app.render does not translate language keys which is why you had the issue in the first place. See how app.render is used in nodebb-widget-essentials along with translator here https://github.com/NodeBB/nodebb-widget-essentials/blob/master/library.js#L194-L198

**Comment by LukeLaupheimer** [comment]
@julianlam Thanks for the suggestion, man! Thing is, going `templates.parse("partials/nwod-roll", renderData, function (html) { callback(html); }` just returns the query (`/roll 5 10`).

I think you guys are just reluctant to let core get changed by some outsider. :P As I've said in previous  posts, I've got a workaround. My plugin works fine. This pull request is a fix to a NodeBB bug. My plugin has nothing to do with it. The only reason I brought it up at all is because in writing it, is how I discovered it.

**Comment by LukeLaupheimer** [comment]
So, yae or nae on fixing this problem with NodeBB?

**Pull request closed without merging**